rfcs/taskforce/0003-Integration-test-infrastructure-spec.md

```diff
@@ -0,0 +1,204 @@
+
+Currently we use Stackdriver as our logging infrastructure for testnets and qa-nets. Using stackdriver, one can set up a pub/sub system to subscribe to a topic ([for more information](https://cloud.google.com/pubsub/docs/quickstart-py-mac)). Log filters can be added to a subscription and so only log messages that match the filter will be published to the topic. Each query defined in the log engine would correspond to a specific topic.
+
+Here's a rough structure of the test executive and how it links different components
+
+Assumption: (as a CI step)
+    1. generate and upload the docker image
+    2. generate genesis proof, genesis ledger with the input runtime configuration using `runtime_genesis_ledger.exe` and upload it to s3
+
+```ocaml
+
+    module type Make_test_intf = functor (Node: Node_intf) :
+        (sig
+            val execute : (module Log_engine : Log_engine_intf) -> Testnet.t -> Runtime_config.t -> bool Deferred.t
+            end)
+
+    module Test_executive
+        (Make_test : Make_test_intf)
+        ( Config : sig
+            val runtime_config : Runtime_config.t
+            val network_config: Network_config.t
+            end) =
+    struct  =
+
+        let execute () =
+            let deploy_config = {daemon_image}
+            in
+            (*deploy the network*)
+            let%bind testnet : Testnet.t = Cloud_network_manager.deploy network_config daemon_config
+            in
+            (*create the log engine based on the testnet configuration*)
+            let module Log_engine = Make_log_engine (struct
+                let testnet = testnet
+                end)
+            in
+            (*Test module for the specific Node implementation*)
+            let module Test = Make_test (Cloud_network_manager.Node)
+            in
+            (*Execute the test*)
+            let%bind result = Test.execute Log_engine testnet runtime_config
+            in
+            (*Teardown the network*)
+            let%map () = Cloud_network_manager.destroy testnet
+            in
+            if not result then failwith "Test failed" else ()
+
+    end
+```
+
+## Outstanding Questions/Concerns
+
+1. We'd still need build profiles because of the constraint constants. This means there would be a docker image uploaded for each build profile.
+2. Would we be running all the tests with snarks? Currently we run integration tests with different proof levels (full | check | none). Some of the tests like the delegation test or bootstrap test that run without snarks would take really long time to complete with proof-level=full.  
```

Comment:
`proof_level` will move into the general config too, so this can be decided per-test.